src/cockpit/superuser.py

```diff
@@ -0,0 +1,215 @@
+        if not options.get('superuser') or self.current == 'root':
+            # superuser not requested, or already superuser?  Next rule.
+            return None
+        elif self.peer:
+            # superuser requested and active?  Return it.
+            return self.peer
+        else:
+            # superuser requested, but not active?  That's an error.
+            raise RoutingError('access-denied')
+
+    # PeerStateListener hooks
+    def peer_state_changed(self, peer: Peer, event: str, exc: Optional[Exception] = None) -> None:
+        logger.debug('Peer %s state changed -> %s', peer.name, event)
+        if event == 'connected':
+            self.current = 'init'
+
+        elif event == 'init':
+            self.peer = peer
+            self.current = peer.name
+            if self.startup is not None:
+                self.startup.success(self)
+                self.startup = None
+
+        elif event == 'closed':
+            self.current = 'none'
+            self.peer = None
+
+            if self.startup is not None:
+                self.startup.failed(self, exc or ConnectionResetError('connection lost'))
+                self.startup = None
+
+    def peer_authorization_request(self, peer: Peer, prompt: str, echo: bool) -> None:
+        if self.startup:
+            self.startup.auth(self, prompt, echo)
+
+    def __init__(self, router: Router, privileged: bool = False):
+        super().__init__(router)
+
+        self.bridges = list(SUPERUSER_BRIDGES)
+        self.peer = None
+        self.startup = None
+
+        if privileged or os.getuid() == 0:
+            self.current = 'root'
+
+    def go(self, startup: SuperuserStartup, name: str) -> None:
+        assert self.current == 'none'
+        assert self.peer is None
+        assert self.startup is None
+
+        args, env = SUPERUSER_BRIDGES[name]
+        startup.peer = Peer(self.router, name, self)
+        startup.peer.spawn(args, env)
+        self.startup = startup
+
+    def init(self, params: Dict[str, object]) -> None:
+        name = params.get('id')
+        if not isinstance(name, str) or name == 'any':
+            name = 'sudo'
+        self.go(ControlMessageStartup(), name)
+
+    # D-Bus methods
+    @bus.Interface.Method(in_types=['s'])
+    async def start(self, name: str) -> None:
+        try:
+            startup = DBusStartup()
+            self.go(startup, name)
+            await startup.wait()
+        except AssertionError as exc:
+            raise bus.BusError('cockpit.Superuser.Error', 'Superuser bridge already running') from exc
+        except KeyError as exc:
+            raise bus.BusError('cockpit.Superuser.Error', 'Unknown superuser bridge type') from exc
+        except OSError as exc:
+            raise bus.BusError('cockpit.Superuser.Error', f'Failed to start peer bridge: {exc}') from exc
+
+    @bus.Interface.Method()
+    def stop(self) -> None:
+        if self.peer is not None:
+            self.peer.close()
```

Comment:
We might need to wait here until the peer has really gone away before sending a reply. JS will call Stop() and then immediately Start() sometimes.

Reply:
Done.